ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
-      List<Datacenter> dataCenters) {
-    ArrayList<Disk> allocatedDisks = new ArrayList<Disk>();
+      Datacenter datacenter, boolean disableRackAwareness) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
+    boolean rackAware = datacenter.isRackAware() && !disableRackAwareness;
+    Set<DataNode> nodesToExclude = new HashSet<>();
+    List<DataNode> dataNodes = new ArrayList<>(datacenter.getDataNodes());
+    for (int i = 0; i < replicaCountPerDatacenter; i++) {
+      List<Disk> diskCandidates =
+          getRandomDiskCandidateSample(dataNodes, nodesToExclude, replicaCapacityInBytes, rackAware, NUM_CHOICES);
+      if (diskCandidates.size() > 0) {
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
+        disksToAllocate.add(diskToAdd);
+        nodesToExclude.add(diskToAdd.getDataNode());
+      }
+    }
+    return disksToAllocate;
+  }
 
-    for (Datacenter datacenter : dataCenters) {
-      int rcpd = replicaCountPerDatacenter;
-      Set<DataNode> nodesToExclude = new HashSet<DataNode>();
-      for (int i = 0; i < rcpd; i++) {
-        DataNode nodeWithMostCapacity = getDataNodeWithMostUnallocatedRawCapacity(datacenter, nodesToExclude);
-        Disk diskWithMostCapacity = getDiskWithMostUnallocatedRawCapacity(nodeWithMostCapacity, replicaCapacityInBytes);
-        allocatedDisks.add(diskWithMostCapacity);
-        nodesToExclude.add(nodeWithMostCapacity);
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}. Retry a non rack-aware allocation
+   * in certain cases described below if {@code bestEffort} is enabled.
+   *
+   * @param replicaCountPerDatacenter how many replicas to attempt to allocate in the datacenter
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param disableRackAwareness if {@code true}, do not attempt a rack-aware allocation
+   * @param bestEffort If {@code bestEffort} is {@code true} and {@code disableRackAwareness} is {@code false},
+   *                   this method will attempt a non rack-aware allocation if it cannot find a rack-aware allocation.
+   * @return a list of {@code replicaCountPerDatacenter} or fewer disks that can be allocated for a new partition in
+   *         the specified datacenter
+   */
+  private List<Disk> allocateDisksForPartitionWithRetry(int replicaCountPerDatacenter, long replicaCapacityInBytes,
+      Datacenter datacenter, boolean disableRackAwareness, boolean bestEffort) {
+
+    List<Disk> disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes,
+        datacenter, disableRackAwareness);
+    if (datacenter.isRackAware() && !disableRackAwareness && (disks.size() < replicaCountPerDatacenter)) {
+      System.err.println("Rack-aware allocation failed for a partition on datacenter:" + datacenter.getName());
+      if (bestEffort) {
```

Comment:
I will do that.  At first, I wanted to print a message if allocation failed and `bestEffort` was `false`, but that probably isn't needed because we print a similar message anyways on line 381.